src/method.c

```diff
@@ -434,28 +391,46 @@ static void jl_method_set_source(jl_method_t *m, jl_code_info_t *src)
                 set_lineno = 1;
             }
         }
-        else if (jl_is_expr(st) && ((jl_expr_t*)st)->head == meta_sym &&
-                 jl_expr_nargs(st) > 1 && jl_exprarg(st, 0) == (jl_value_t*)nospecialize_sym) {
-            for (size_t j=1; j < jl_expr_nargs(st); j++) {
-                jl_value_t *aj = jl_exprarg(st, j);
-                if (jl_is_slot(aj)) {
-                    int sn = (int)jl_slot_number(aj) - 2;
-                    if (sn >= 0) {  // @nospecialize on self is valid but currently ignored
-                        if (sn > (m->nargs - 2)) {
-                            jl_error("@nospecialize annotation applied to a non-argument");
-                        }
-                        else if (sn >= sizeof(m->nospecialize) * 8) {
-                            jl_printf(JL_STDERR,
-                                      "WARNING: @nospecialize annotation only supported on the first %d arguments.\n",
-                                      (int)(sizeof(m->nospecialize) * 8));
-                        }
-                        else {
-                            m->nospecialize |= (1 << sn);
+        else if (jl_is_expr(st) && ((jl_expr_t*)st)->head == meta_sym) {
+            if (jl_expr_nargs(st) > 1 && jl_exprarg(st, 0) == (jl_value_t*)nospecialize_sym) {
+                for (size_t j=1; j < jl_expr_nargs(st); j++) {
+                    jl_value_t *aj = jl_exprarg(st, j);
+                    if (jl_is_slot(aj)) {
+                        int sn = (int)jl_slot_number(aj) - 2;
+                        if (sn >= 0) {  // @nospecialize on self is valid but currently ignored
+                            if (sn > (m->nargs - 2)) {
+                                jl_error("@nospecialize annotation applied to a non-argument");
+                            }
+                            else if (sn >= sizeof(m->nospecialize) * 8) {
+                                jl_printf(JL_STDERR,
+                                          "WARNING: @nospecialize annotation only supported on the first %d arguments.\n",
+                                          (int)(sizeof(m->nospecialize) * 8));
+                            }
+                            else {
+                                m->nospecialize |= (1 << sn);
+                            }
                         }
                     }
                 }
+                st = jl_nothing;
+            }
+            else if (jl_expr_nargs(st) == 2 && jl_exprarg(st, 0) == (jl_value_t*)generator_sym) {
+                m->generator = NULL;
+                JL_TRY {
+                    m->generator = jl_toplevel_eval(m->module, jl_exprarg(st, 1));
```

Comment:
I feel like our `deserialize` code might be assuming that this function is effect-free / non-reentrant.

Reply:
Also, what sorts of code is this expecting to potentially encounter? The method-world-counter code, and the multithreaded-toplevel-lock code also strongly assume / require that this function must be effect-free and non-reentrant (both are currently quite bad at enforcing it, but that's besides the point).

Reply:
The expression it expects is `(new (core GeneratedFunctionStub) globalvar constant constant constant constant)`, so all it does is read two global vars and allocate a struct. I could do that manually instead of calling `eval` if you think it's necessary.